lumeer-storage/lumeer-storage-api/src/main/java/io/lumeer/storage/api/query/SearchQueryStem.java

```diff
@@ -0,0 +1,158 @@
+/*
+ * Lumeer: Modern Data Definition and Processing Platform
+ *
+ * Copyright (C) since 2017 Answer Institute, s.r.o. and/or its affiliates.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package io.lumeer.storage.api.query;
+
+import io.lumeer.api.model.ConditionType;
+import io.lumeer.api.model.QueryStem;
+import io.lumeer.storage.api.filter.AttributeFilter;
+
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+public class SearchQueryStem {
```

Comment:
If we want to have the whole `Query` immutable, we need to make stems immutable as well by returning unmodifiable collections (and annotate it just for record). If we need to modify the query directly somewhere and it will take some time to rewrite it, we can skip this whole immutability thing altogether and remove all traces of it.